File: QualDS/integrate_fire.c
#include <stdio.h>

int main()
{
  int nr_steps = 1000;
  double dt = 0.001;
  double t[nr_steps];
  int i;
  for (i=0;i<nr_steps;i++){
    t[i] = (double) i*dt;
  }
  double v[nr_steps];
  v[0] = 0.0;
  for(i=1;i<nr_steps;i++){
    v[i] = 0.0;
  }
  double tau = 0.1;
  double I = 2.0;
  double thresh = 1.0;
  double reset = -1.0;
  for (i=0;i<nr_steps;i++){
    double dv_dt = -(v[i]-I)/tau;
    double vv = v[i] + dv_dt*dt;
    if (vv > thresh){
      v[i+1] = reset;
    }
    else {
      v[i+1] = vv;
    }
  }
  for(i=0;i<nr_steps;i++){
    printf("t = %f, v = %f\n", t[i], v[i]);
  }
  return 0;
}

File: QualDS/integrate_fire.py
import numpy as np
from scipy.integrate import odeint
import matplotlib.pyplot as plt

nsteps = 1000

t = np.linspace(0, 100, nsteps)

I = 2.
tau = 0.1
thresh = 1.
reset = -1.
v0 = 0.

def dV(vt,ti):
    return -(vt-I)/tau

v = np.zeros(nsteps)
dt = 0.001
for i in range(len(t)-1):
    dvdt = dV(v[i],t[i],)
    vv = v[i] + dvdt*dt
    if vv > thresh:
        v[i + 1] = reset
    else:
        v[i + 1] = vv
    print(v[i+1])

plt.figure()
plt.plot(t, v)

plt.figure()
plt.plot(t, np.exp(t/tau)*(I-v))
plt.show()
